lib/spack/spack/ci.py

```diff
@@ -2020,50 +2035,66 @@ def process_command(name, commands, repro_dir, run=True, exit_on_failure=True):
 
     Returns: the exit code from processing the command
     """
-    tty.debug("spack {0} arguments: {1}".format(name, commands))
 
+    tty.debug(f"spack {name} arguments: {commands}")
     if len(commands) == 0 or isinstance(commands[0], str):
         commands = [commands]
-
-    # Create a string [command 1] && [command 2] && ... && [command n] with commands
-    # quoted using double quotes.
-    args_to_string = lambda args: " ".join('"{}"'.format(arg) for arg in args)
-    full_command = " \n ".join(map(args_to_string, commands))
-
-    # Write the command to a shell script
-    script = "{0}.sh".format(name)
+    ext = "ps1" if IS_WINDOWS else "sh"
+    verbose_commands = "Set-PSDebug -Trace 2\n" if IS_WINDOWS else "set -x\n"
+    shebang = "" if IS_WINDOWS else "#!/bin/sh\n\n"
+    exit_error = '$ErrorActionPreference = "Stop"\n' if IS_WINDOWS else "set -e\n"
+    # ErrorActionPreference will handle PWSH commandlets (Spack calls),
+    # but we need to handle EXEs (git, etc) ourselves
+    catch_exe_failure = (
+        """
+if ($LASTEXITCODE -ne 0){
+    throw "Command {} has failed"
+}
+"""
+        if IS_WINDOWS
+        else ""
+    )
+    # Create a string [command 1] \n [command 2] \n ... \n [command n] with
+    # commands composed into a platform dependent shell script, pwsh on Windows,
+    args_to_string = lambda args: f'{" ".join(args)}\n{catch_exe_failure.format(" ".join(args))}'
+    full_command = "\n".join(map(args_to_string, commands))
+    # Write the command to a python script
+    script = f"{name}.{ext}"
     with open(script, "w") as fd:
-        fd.write("#!/bin/sh\n\n")
-        fd.write("\n# spack {0} command\n".format(name))
-        if exit_on_failure:
-            fd.write("set -e\n")
+        fd.write(shebang)
+        fd.write(f"\n# spack {name} command\n")
+        fd.write(f"{exit_error}")
         if os.environ.get("SPACK_VERBOSE_SCRIPT"):
-            fd.write("set -x\n")
+            fd.write(verbose_commands)
         fd.write(full_command)
         fd.write("\n")
 
-    st = os.stat(script)
-    os.chmod(script, st.st_mode | stat.S_IEXEC)
+    if not IS_WINDOWS:
+        st = os.stat(script)
+        os.chmod(script, st.st_mode | stat.S_IEXEC)
 
     copy_path = os.path.join(repro_dir, script)
     shutil.copyfile(script, copy_path)
-    st = os.stat(copy_path)
-    os.chmod(copy_path, st.st_mode | stat.S_IEXEC)
+    if not IS_WINDOWS:
+        st = os.stat(copy_path)
+        os.chmod(copy_path, st.st_mode | stat.S_IEXEC)
```

Comment:
(question) Since the script is being passed to an interpreter, does the `/bin/sh` version require this?

Reply:
I think the permissions change is required for `copy_path` (since it's the script that persists and might be used elsewhere by another actor) but is not for `script`, since, as you point out, it's being driven by `/bin/sh` anyway.